nntrainer/models/neuralnet.cpp

```diff
@@ -168,90 +168,8 @@ int NeuralNetwork::initialize() {
-    bool first = idx == 0;
-    auto &lnode = model_graph.getSortedLayerNode(idx);
-    Layer &l = *lnode.getObject();
-    ml_logd("layer name : %s", l.getName().c_str());
-    std::string cur_type;
-    if (l.getType() == TimeDistLayer::type) {
-      cur_type = dynamic_cast<TimeDistLayer &>(l).getDistLayerType();
-    } else {
-      cur_type = l.getType();
-    }
-
-    /**
-     * Set input dimension for all the layers.
-     * For input layer, as input dimension is known, set input tensor.
-     */
-    if (!first) {
-      std::string l_pre_type =
-        model_graph.getSortedLayerNode(idx - 1).getObject()->getType();
-      if (l_pre_type == TimeDistLayer::type) {
-        l_pre_type = std::dynamic_pointer_cast<TimeDistLayer>(
-                       model_graph.getSortedLayerNode(idx - 1).getObject())
-                       ->getDistLayerType();
-      }
-      if (istrequal(l_pre_type, ActivationLayer::type) &&
-          istrequal(cur_type, ActivationLayer::type)) {
-        ml_loge("double activation is not allowed");
-        return ML_ERROR_INVALID_PARAMETER;
-      }
-
-      for (unsigned int i = 0; i < l.input_layers.size(); ++i) {
-        Layer &in_layer =
-          *model_graph.getLayerNode(l.input_layers[i]).getObject();
-
-        unsigned int location = 0;
-        for (unsigned int j = 0; j < in_layer.output_layers.size(); ++j) {
-          if (in_layer.output_layers[j] == l.getName()) {
-            location = j;
-            break;
-          }
-        }
-
-        l.setInputDimension(in_layer.getOutputDimension()[location], i);
-      }
-    }
-
-    /**
-     * Initialize all the layers, allocate output tensors for each layer
-     * and add optimizer related weights for the layer
-     */
-    status = l.initialize(*manager);
-    NN_RETURN_STATUS();
-
-    REGISTER_EVENT(l.getName(), lnode.event_key)
-
-    auto &in_out = manager->trackLayerOutputs(
-      cur_type, l.getName(), l.getOutputDimension(), l.getInputDimension());
-    l.setOutputBuffers(in_out);
-
-    /** Connect the output of the previous layers with the input of the current
-     * layer */
-    if (!first) {
-      for (unsigned int i = 0; i < l.input_layers.size(); ++i) {
-        Layer &in_layer =
-          *model_graph.getLayerNode(l.input_layers[i]).getObject();
-
-        unsigned int location = 0;
-        for (unsigned int j = 0; j < in_layer.output_layers.size(); ++j) {
-          if (in_layer.output_layers[j] == l.getName()) {
-            location = j;
-            break;
-          }
-        }
-
-        l.net_input[i] = model_graph.getLayerNode(l.input_layers[i])
-                           .getObject()
-                           ->net_hidden[location];
-      }
-    } else {
-      auto &in_out = manager->trackLayerInputs(
-        cur_type, l.getName(), l.getInputDimension(), l.getOutputDimension());
-      l.setInputBuffers(in_out);
-    }
-  }
+  status = model_graph.initialize(manager);
```

Comment:
:100:

Reply:
with this, does manager fit better inside graph than in model class?

Reply:
I think so

Reply:
then let's do so in a later PR.